Use the QUICKLIST encoding to dump the list payload to compatible with Redis >= 4

AntiTopQuark opened the pull request:
Close #2253

In order to support the compatibility of old versions of redis, 

when using the dump command, `QUICKLIST`encoding is used instead of `QUICKLIST2`for LIST encoding.

![](https://github.com/apache/kvrocks/assets/32798045/d81c2623-f81a-4c2a-b64f-66297d40291e)

AntiTopQuark committed:
fix: dumped payload can't use to restore in redis v6.2

AntiTopQuark committed:
Merge branch 'unstable' into unstable

mapleFU commented on src/storage/rdb.cc:
```diff
@@ -1005,3 +1000,37 @@ Status RDB::rdbSaveBinaryDoubleValue(double val) {
+  const size_t ziplist_size = zlHeaderSize + zlEndSize + elem.length() +
+                              ZipList::zipStorePrevEntryLength(nullptr, prevlen) +
+                              ZipList::zipStoreEntryEncoding(nullptr, elem.length());
+  auto zl = new unsigned char[ziplist_size];
```
why new explicitly here?

AntiTopQuark replied:
I have changed the code from using `new` directly to using `std::make_unique` for memory management. 
Thank you for your review and suggestions!

AntiTopQuark committed:
update

AntiTopQuark committed:
update

PragmaTwice commented:
Could you write some test cases for your patch? Thank you!

AntiTopQuark committed:
update

AntiTopQuark commented:
> Could you write some test cases for your patch? Thank you!

Here I just changed a coding method. The previous tests can still be used. I have now added some corner cases.

PragmaTwice commented:
> > Could you write some test cases for your patch? Thank you!
> 
> Here I just changed a coding method. The previous tests can still be used. I have now added some corner cases.

Generally, if we fixed a bug in a patch, we need to add a test case, that should fail without the patch, and succeed after the patch is applied.

AntiTopQuark committed:
update

AntiTopQuark committed:
update

git-hulk commented on src/storage/rdb_ziplist.h:
```diff
@@ -41,4 +53,4 @@ class ZipList {
   void setPreEntryLen(uint32_t len) { pre_entry_len_ = len; }
   static uint32_t getEncodedLengthSize(uint32_t len);
-};
+};
```
```suggestion
};

```

Keep the newline at the end of file.

AntiTopQuark replied:
done

git-hulk commented on src/storage/rdb_ziplist.cc:
```diff
@@ -152,3 +148,65 @@ Status ZipList::peekOK(size_t n) {
+  return 1 + sizeof(uint32_t);
+}
+
+uint32_t ZipList::ZipStorePrevEntryLength(unsigned char *p, unsigned int len) {
+  if (p == nullptr) {
+    return (len < ZipListBigLen) ? 1 : sizeof(uint32_t) + 1;
+  } else {
+    if (len < ZipListBigLen) {
+      p[0] = len;
+      return 1;
+    } else {
+      return ZipStorePrevEntryLengthLarge(p, len);
+    }
+  }
+}
```
```suggestion
uint32_t ZipList::ZipStorePrevEntryLength(unsigned char *p, unsigned int len) {
  if (p == nullptr) {
    return (len < ZipListBigLen) ? 1 : sizeof(uint32_t) + 1;
  }
  if (len < ZipListBigLen) {
    p[0] = len;
    return 1;
  }
  return ZipStorePrevEntryLengthLarge(p, len);
}
```

AntiTopQuark replied:
done

git-hulk commented:
@AntiTopQuark Thanks for your efforts.

mapleFU commented on src/storage/rdb.cc:
```diff
@@ -1005,3 +1000,35 @@ Status RDB::rdbSaveBinaryDoubleValue(double val) {
+  const size_t ziplist_size = zlHeaderSize + zlEndSize + elem.length() +
+                              ZipList::ZipStorePrevEntryLength(nullptr, prevlen) +
+                              ZipList::ZipStoreEntryEncoding(nullptr, elem.length());
+  auto zl = std::make_unique<unsigned char[]>(ziplist_size);
```
would an std::string better here? which also avoid the copying in L:1027

AntiTopQuark replied:
thank you very much, done

AntiTopQuark committed:
update

AntiTopQuark committed:
Merge branch 'unstable' into unstable

git-hulk commented:
I can confirm it also works in Redis 4:

```
127.0.0.1:6379> RESTORE list 0 "\x0e\x04\x0e\x0e\x00\x00\x00\n\x00\x00\x00\x01\x00\x00\x01d\xff\x0e\x0e\x00\x00\x00\n\x00\x00\x00\x01\x00\x00\x01c\xff\x0e\x0e\x00\x00\x00\n\x00\x00\x00\x01\x00\x00\x01b\xff\x0e\x0e\x00\x00\x00\n\x00\x00\x00\x01\x00\x00\x01a\xff\x06\x003\x06\xd4\x86\x93\xfbvB"
OK
127.0.0.1:6379> lrange list 0 -1
1) "d"
2) "c"
3) "b"
4) "a"
```

mapleFU commented on src/storage/rdb.cc:
```diff
@@ -1005,3 +1000,35 @@ Status RDB::rdbSaveBinaryDoubleValue(double val) {
+  const size_t ziplist_size = zlHeaderSize + zlEndSize + elem.length() +
+                              ZipList::ZipStorePrevEntryLength(nullptr, prevlen) +
+                              ZipList::ZipStoreEntryEncoding(nullptr, elem.length());
+  auto zl_string = std::string(ziplist_size, 0);
```
Why the size would be 0? Isn't it same as `ziplist_size`?

git-hulk committed:
Merge branch 'unstable' into unstable

git-hulk committed:
Merge branch 'unstable' into unstable

mapleFU commented on src/storage/rdb_ziplist.cc:
```diff
@@ -152,3 +148,63 @@ Status ZipList::peekOK(size_t n) {
+  (*((uint32_t *)((zl) + sizeof(uint32_t)))) = value;
+}
+void ZipList::SetZipListLength(unsigned char *zl, uint16_t value) {
+  (*((uint16_t *)((zl) + sizeof(uint32_t) * 2))) = value;
```
Would you mind use memcpy? Since this might be a type puning here

AntiTopQuark replied:
done

mapleFU commented on src/storage/rdb_ziplist.cc:
```diff
@@ -152,3 +148,63 @@ Status ZipList::peekOK(size_t n) {
+
+void ZipList::SetZipListBytes(unsigned char *zl, uint32_t value) { (*((uint32_t *)(zl))) = value; }
+void ZipList::SetZipListTailOffset(unsigned char *zl, uint32_t value) {
+  (*((uint32_t *)((zl) + sizeof(uint32_t)))) = value;
```
ditto, would you mind use memcpy here?

AntiTopQuark replied:
done

mapleFU commented on src/storage/rdb_ziplist.cc:
```diff
@@ -152,3 +148,63 @@ Status ZipList::peekOK(size_t n) {
+  return len;
+}
+
+void ZipList::SetZipListBytes(unsigned char *zl, uint32_t value) { (*((uint32_t *)(zl))) = value; }
```
Would you mind use memcpy here?

AntiTopQuark replied:
done

mapleFU commented on src/storage/rdb.cc:
```diff
@@ -1005,3 +1000,35 @@ Status RDB::rdbSaveBinaryDoubleValue(double val) {
+  auto zl_ptr = reinterpret_cast<unsigned char *>(&zl_string[0]);
+
+  // set ziplist header
+  ZipList::SetZipListBytes(zl_ptr, intrev32ifbe(ziplist_size));
```
`ziplist_size` is not a uint32, right? Would we check and cast it to u32?

AntiTopQuark replied:
done

mapleFU commented on src/storage/rdb_ziplist.cc:
```diff
@@ -152,3 +148,63 @@ Status ZipList::peekOK(size_t n) {
+  memcpy(p, buf, len);
+  return len;
+}
+
```
Besides, would we also applying length check for these set below?

AntiTopQuark replied:
done

mapleFU commented on src/storage/rdb_ziplist.cc:
```diff
@@ -152,3 +148,63 @@ Status ZipList::peekOK(size_t n) {
+  if (p != nullptr) {
+    p[0] = ZipListBigLen;
+    u32 = len;
+    memcpy(p + 1, &u32, sizeof(u32));
+    memrev32ifbe(p + 1);
```
Just curious why memrev after memcpy? Wouldn't that be faster to rev the register before writing into memory?

AntiTopQuark replied:
Currently, it seems that Redis is implemented in the same way. I suspect it is to avoid the risks associated with operating directly on the original data.

mapleFU reviewed (commented):
General LGTM. Remaining problems are about code styles

AntiTopQuark committed:
update

AntiTopQuark committed:
Merge branch 'unstable' of github.com:AntiTopQuark/kvrocks into unstable

AntiTopQuark committed:
update

mapleFU commented on src/storage/rdb.cc:
```diff
@@ -1005,3 +1000,36 @@ Status RDB::rdbSaveBinaryDoubleValue(double val) {
+  ZipList::SetZipListLength(zl_ptr, ziplist_size, 1);
+  zl_ptr[ziplist_size - 1] = zlEnd;
+
+  if (ziplist_size > 0) {
```
I'm reading the logic here, may I ask that when would `ziplist_size == 0`? Since if `ziplist_size == 0`, the code above would hit memory issue? If it cannot be 0, we should remove this branch?

AntiTopQuark replied:
oh, you are right. done～

AntiTopQuark committed:
update

mapleFU commented:
I didn't dive into logics of RDB carefully, generally this LGTM

git-hulk commented:
@AntiTopQuark Need to reformat codes.

AntiTopQuark committed:
update

git-hulk committed:
Merge branch 'unstable' into unstable
